src/app/berkeley_migration_verifier/berkeley_migration_verifier.ml

```diff
@@ -0,0 +1,516 @@
+      Test.of_check check ~name:"No orphaned blocks in migrated schema" ~idx:8
+      |> Test.eval ;
+
+      let%bind check =
+        compare_hashes
+          ~fetch_data_sql:(fun db ->
+            Sql.Mainnet.ledger_hashes_no_orphaned db end_global_slot )
+          ~find_element_sql:(fun hash ->
+            query_migrated_db ~f:(fun db ->
+                Sql.Berkeley.find_block_by_ledger_hash db hash ) )
+          ~name:"ledger_hashes"
+      in
+
+      Test.of_check check ~name:"No orphaned ledger hashes in migrated schema"
+        ~idx:9
+      |> Test.eval ;
+
+      let%bind expected_hashes =
+        query_mainnet_db ~f:(fun db ->
+            Sql.Common.block_state_hashes db end_global_slot )
+      in
+      let%bind actual_hashes =
+        query_migrated_db ~f:(fun db ->
+            Sql.Common.block_state_hashes db end_global_slot )
+      in
+
+      List.map expected_hashes ~f:(fun (expected_child, expected_parent) ->
+          if
+            List.exists actual_hashes ~f:(fun (actual_child, actual_parent) ->
+                String.equal expected_child actual_child
+                && String.equal expected_parent actual_parent )
+          then Check.ok
+          else
+            Check.err
+              (sprintf
+                 "Relation between blocks is skewed. Cannot find original \
+                  subchain '%s' -> '%s' in migrated database"
+                 expected_child expected_parent ) )
+      |> Check.combine
+      |> Test.of_check ~name:"Block relation parent -> child preserved" ~idx:10
+      |> Test.eval ;
+      Deferred.unit
+
+let get_forked_blockchain uri fork_point =
+  match Caqti_async.connect_pool ~max_size:128 uri with
+  | Error _ ->
+      failwithf "Failed to create Caqti pools for Postgresql to %s"
+        (Uri.to_string uri) ()
+  | Ok pool ->
+      let query_db = Mina_caqti.query pool in
+      query_db ~f:(fun db ->
+          Sql.Berkeley.Block_info.forked_blockchain db fork_point )
+
+let check_forked_chain ~migrated_archive_uri ~fork_state_hash =
+  let%bind forked_blockchain =
+    get_forked_blockchain (Uri.of_string migrated_archive_uri) fork_state_hash
+  in
+
+  match forked_blockchain with
+  | [] ->
+      Deferred.return (Check.skipped "Forked state hash does not exist")
+  | [ _fork_point ] ->
+      Deferred.return (Check.skipped "Forked blockchain is empty")
+  | fork_point :: fork_chain ->
+      Deferred.return
+        ( List.mapi fork_chain ~f:(fun idx block ->
+              let idx_64 = Int64.of_int idx in
+              let protocol_version_check =
+                if Int.( <> ) block.protocol_version_id 2 then
+                  Check.err
+                    (sprintf
+                       "block with id (%d) has unexpected protocol version"
+                       block.id )
+                else Check.ok
+              in
+              let global_slot_since_hardfork_check =
+                if Int64.( = ) block.global_slot_since_hard_fork idx_64 then
```

Comment:
This comparison also makes little sense. And it could possibly fail if we have many pending blocks on the migrated database.